prov/util/src/util_srx.c

```diff
@@ -0,0 +1,1155 @@
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <stdlib.h>
+
+#include <ofi_enosys.h>
+#include "ofi_iov.h"
+#include <ofi_util.h>
+
+static struct util_rx_entry *util_alloc_rx_entry(struct util_srx_ctx *srx)
+{
+	return (struct util_rx_entry *) ofi_buf_alloc(srx->rx_pool);
+}
+
+static inline struct iovec *util_srx_iov(struct util_rx_entry *rx_entry)
+{
+	return (struct iovec *) ((char *) rx_entry + sizeof(*rx_entry));
+}
+
+static inline void **util_srx_desc(struct util_srx_ctx *srx,
+				   struct util_rx_entry *rx_entry)
+{
+	return (void **) ((char *) util_srx_iov(rx_entry) +
+			(sizeof(struct iovec) * srx->iov_limit));
+}
+
+static void util_init_rx_entry(struct util_rx_entry *entry,
+			       const struct iovec *iov, void **desc,
+			       size_t count, fi_addr_t addr, void *context,
+			       uint64_t tag, uint64_t flags)
+{
+	memcpy(entry->peer_entry.iov, iov, sizeof(*iov) * count);
+	if (desc)
+		memcpy(entry->peer_entry.desc, desc, sizeof(*desc) * count);
+	else
+		memset(entry->peer_entry.desc, 0, sizeof(*desc) * count);
+
+	entry->peer_entry.count = count;
+	entry->peer_entry.addr = addr;
+	entry->peer_entry.context = context;
+	entry->peer_entry.tag = tag;
+	entry->peer_entry.flags = flags;
+}
+
+static struct util_rx_entry *util_get_recv_entry(struct util_srx_ctx *srx,
+		const struct iovec *iov, void **desc, size_t count,
+		fi_addr_t addr, void *context, uint64_t tag, uint64_t ignore,
+		uint64_t flags)
+{
+	struct util_rx_entry *entry;
+
+	entry = util_alloc_rx_entry(srx);
+	if (!entry)
+		return NULL;
+
+	util_init_rx_entry(entry, iov, desc, count, addr, context, tag, flags);
+
+	entry->peer_entry.owner_context = NULL;
+
+	entry->multi_recv_ref = 0;
+	entry->ignore = ignore;
+	entry->seq_no = srx->rx_seq_no++;
+
+	return entry;
+}
+
+static struct util_rx_entry *util_init_unexp(struct util_srx_ctx *srx,
+			fi_addr_t addr, uint64_t size, uint64_t tag)
+{
+	struct util_rx_entry *util_entry;
+
+	util_entry = util_alloc_rx_entry(srx);
+	if (!util_entry)
+		return NULL;
+
+	util_entry->peer_entry.owner_context = NULL;
+	util_entry->peer_entry.size = size;
+	util_entry->peer_entry.addr = addr;
+	util_entry->peer_entry.tag = tag;
+
```

Comment:
Should we also set `util_entry->peer_entry.desc = NULL` here? Not a bug in this PR, but there was some bug in my development which was partly due to using an unexp entry with garbage (non-null) desc.

Reply:
We can if you like but I do think it's a waste since the unexpected message isn't attached to a receive yet and shouldn't have any desc attached. The peer should never access the desc. It is initialized correctly initialized when it is actually matched to a posted receive, before the update mr function. If it is fixing an issue in efa I think that means efa is incorrectly accessing the desc when it shouldn't

Reply:
Makes sense. No bother then !